core/src/accounts_hash_verifier.rs

```diff
@@ -155,17 +179,53 @@ impl AccountsHashVerifier {
             }
         }
 
+        cluster_info.push_accounts_hashes(hashes.clone());
+    }
+
+    fn handle_snapshot(
+        accounts_package: AccountsPackage,
+        pending_snapshot_package: Option<&PendingSnapshotPackage>,
+        snapshot_config: Option<&SnapshotConfig>,
+    ) {
         if let Some(snapshot_config) = snapshot_config {
-            if snapshot_package.block_height % snapshot_config.full_snapshot_archive_interval_slots
-                == 0
-            {
-                if let Some(pending_snapshot_package) = pending_snapshot_package {
+            if let Some(pending_snapshot_package) = pending_snapshot_package {
+                if accounts_package.block_height
+                    % snapshot_config.full_snapshot_archive_interval_slots
+                    == 0
+                {
+                    let snapshot_package = SnapshotPackage::new(
+                        accounts_package,
+                        snapshot_config.snapshot_package_output_path.clone(),
+                        snapshot_config.archive_format,
+                        None,
+                    );
+
                     *pending_snapshot_package.lock().unwrap() = Some(snapshot_package);
-                }
-            }
-        }
+                } else if accounts_package.block_height
+                    % snapshot_config.incremental_snapshot_archive_interval_slots
+                    == 0
+                {
+                    let last_full_snapshot_slot =
+                        snapshot_utils::get_highest_full_snapshot_archive_slot(
```

Comment:
Is there a race here if the packaging service is *just* about to finish packaging the next full snapshot, but we read off the old last full snapshot?

Reply:
Yeah... 

Since SnapshotPackagerService is the only place that knows if `snapshot_utils::archive_snapshot_package()` succeeds or not, I think I need to use that result to pass back info to AccountsHashVerifier to get the _true_ last full snapshot slot. This will also be useful for AccountsBackgroundService, because that's where `clean_accounts()` gets called, and it needs the last full snapshot slot as well.

Back to AccountsHashVerifier here, luckily this race doesn't produce broken incremental snapshots, just ones that are too big (because the base slot could've been higher).

Reply:
Fixed!

Reply:
@carllin I fixed this by adding in a shared `last_full_snapshot_slot` to `SnapshotConfig`:

https://github.com/solana-labs/solana/blob/a2491c73865a19e596162b5cc05d67cece943251/runtime/src/snapshot_config.rs#L33-L37

`SnapshotPackagerService` updates this value:

https://github.com/solana-labs/solana/blob/a2491c73865a19e596162b5cc05d67cece943251/core/src/snapshot_packager_service.rs#L48-L56

and `AccountsHashVerifier` reads this value:

https://github.com/solana-labs/solana/blob/a2491c73865a19e596162b5cc05d67cece943251/core/src/accounts_hash_verifier.rs#L223-L224

I'll also end up using this in `AccountsBackgroundService` as well, where `clean_accounts()` is called.

Reply:
Hmm this doesn't seem quite right because we could have an incremental snapshot that arrives before the latest full snapshot `B` is finished. This incremental snapshot will mark the latest full snapshot is some `A < B`, but has already filtered out the storages between `A` and `B`.

I think maybe the right thing to do is use modular arithmetic on the full snapshot interval to figure out what the *expected* last snapshot should be  for an incremental snapshot, and just use that?